ethcore/wasm/src/env.rs

```diff
@@ -14,183 +14,276 @@
+			max_memory: max_memory,
+			.. Default::default()
+		}
+	}
+
+	/// Returns memory that was instantiated during the contract module
+	/// start. If contract does not use memory at all, the dummy memory of length (0, 0)
+	/// will be created instead. So this method always returns memory instance
+	/// unless errored.
+	pub fn memory_ref(&self) -> Result<MemoryRef, Error> {
+		{
+			let mut mem_ref = self.memory.borrow_mut();
+			if mem_ref.is_none() { *mem_ref = Some(MemoryInstance::alloc(0, Some(0))?); }
+		}
+
+		Ok(self.memory.borrow().clone().expect("it is either existed or was created as (0, 0) above; qed"))
+	}
+
+	/// Returns memory size module initially requested
+	pub fn memory_size(&self) -> Result<u32, Error> {
+		Ok(self.memory_ref()?.size())
 	}
 }
+
+impl wasmi::ModuleImportResolver for ImportResolver {
+	// todo: check against `_signature` ?
+	fn resolve_func(&self, field_name: &str, _signature: &Signature) -> Result<FuncRef, Error> {
+		let func_ref = match field_name {
+			"storage_read" => host(signatures::STORAGE_READ, ids::STORAGE_READ_FUNC),
+			"storage_write" => host(signatures::STORAGE_WRITE, ids::STORAGE_WRITE_FUNC),
+			"ret" => host(signatures::RET, ids::RET_FUNC),
+			"gas" => host(signatures::GAS, ids::GAS_FUNC),
+			"input_length" => host(signatures::INPUT_LENGTH, ids::INPUT_LENGTH_FUNC),
+			"fetch_input" => host(signatures::FETCH_INPUT, ids::FETCH_INPUT_FUNC),
+			"panic" => host(signatures::PANIC, ids::PANIC_FUNC),
+			"debug" => host(signatures::DEBUG, ids::DEBUG_FUNC),
+			"ccall" => host(signatures::CCALL, ids::CCALL_FUNC),
+			"dcall" => host(signatures::DCALL, ids::DCALL_FUNC),
+			"scall" => host(signatures::SCALL, ids::SCALL_FUNC),
+			"value" => host(signatures::VALUE, ids::VALUE_FUNC),
+			"create" => host(signatures::CREATE, ids::CREATE_FUNC),
+			"suicide" => host(signatures::SUICIDE, ids::SUICIDE_FUNC),
+			"blockhash" => host(signatures::BLOCKHASH, ids::BLOCKHASH_FUNC),
+			"blocknumber" => host(signatures::BLOCKNUMBER, ids::BLOCKNUMBER_FUNC),
+			"coinbase" => host(signatures::COINBASE, ids::COINBASE_FUNC),
+			"difficulty" => host(signatures::DIFFICULTY, ids::DIFFICULTY_FUNC),
+			"gaslimit" => host(signatures::GASLIMIT, ids::GASLIMIT_FUNC),
+			"timestamp" => host(signatures::TIMESTAMP, ids::TIMESTAMP_FUNC),
+			"address" => host(signatures::ADDRESS, ids::ADDRESS_FUNC),
+			"sender" => host(signatures::SENDER, ids::SENDER_FUNC),
+			"origin" => host(signatures::ORIGIN, ids::ORIGIN_FUNC),
+			"elog" => host(signatures::ELOG, ids::ELOG_FUNC),
+			_ => {
+				return Err(wasmi::Error::Instantiation(
+					format!("Export {} not found", field_name),
+				))
+			}
+		};
+
+		Ok(func_ref)
+	}
+
+	fn resolve_memory(
+		&self,
+		field_name: &str,
+		descriptor: &MemoryDescriptor,
+	) -> Result<MemoryRef, Error> {
+		if field_name == "memory" {
+			if descriptor.initial() >= self.max_memory ||
+				(descriptor.maximum().map_or(false, |m| m < descriptor.initial() || m > self.max_memory))
```

Comment:
This is an interesting check.

If I understand the intent correctly, it checks whether `initial` or `maximum` number of memory pages requested by the module is greater than some limit. In case if `maximum` isn't provided then this check is passed and runtime proceeds to execution.

If that's correct, then I don't see the point. What stops the contract writer to not set `maximum` limit on memory and then allocate amount of memory larger than the limit (don't we have the control on `grow_memory` op, do we?).

And as I remember, don't we agreed that we don't want to restrict the upper memory limit by any means except the gas price?